File: web/lib/types/academic.ts
/**
 * Manuel TypeScript tipleri — `/api/v2/teacher/academic` +
 * `/api/v2/teacher/grade-advance` + `/api/v2/teacher/csv` için.
 *
 * Pydantic şemalarıyla (`app/routes/api_v2/schemas/academic.py`) birebir aynı.
 */

export type PhaseKind = "regular" | "winter_break" | "summer_camp" | "exam_prep";
export type ExamTarget = "lgs" | "yks" | "none";
export type Track = "sayisal" | "ea" | "sozel" | "dil";
export type GraduateMode = "full_time" | "dershane";

// =============================================================================
// Academic year + phase
// =============================================================================

export interface PhaseItem {
  id: number;
  name: string;
  start_date: string; // YYYY-MM-DD
  end_date: string;
  kind: PhaseKind;
  kind_label: string;
  kind_badge: string;
  notes: string | null;
  capacity_multiplier: number;
  is_no_school: boolean;
}

export interface AcademicYearListItem {
  id: number;
  name: string;
  start_year: number | null;
  exam_target: ExamTarget;
  exam_label: string;
  is_active: boolean;
  phase_count: number;
  student_count: number;
  created_at: string;
}

export interface AcademicYearListResponse {
  items: AcademicYearListItem[];
  current_start_year: number;
}

export interface AcademicYearAssignedStudent {
  student_id: number;
  full_name: string;
  grade_level: number | null;
  is_graduate: boolean;
}

export interface AcademicYearDetailResponse {
  id: number;
  name: string;
  start_year: number | null;
  exam_target: ExamTarget;
  exam_label: string;
  is_active: boolean;
  created_at: string;
  phases: PhaseItem[];
  assigned_students: AcademicYearAssignedStudent[];
}

export interface AcademicYearCreateBody {
  start_year: number;
}

export interface AcademicYearPatchBody {
  name?: string | null;
  start_year?: number | null;
  exam_target?: ExamTarget | null;
  is_active?: boolean | null;
}

export interface PhaseCreateBody {
  name: string;
  start_date: string;
  end_date: string;
  kind?: PhaseKind;
  notes?: string | null;
}

export interface PhasePatchBody {
  name?: string | null;
  start_date?: string | null;
  end_date?: string | null;
  kind?: PhaseKind | null;
  notes?: string | null;
}

export interface AcademicYearAssignBody {
  student_ids: number[];
}

export interface AcademicYearAssignResult {
  assigned_count: number;
  removed_count: number;
  unchanged_count: number;
}

export interface AcademicYearListChoiceItem {
  start_year: number;
  name: string;
  label: string;
  exists: boolean;
}

export interface AcademicYearChoicesResponse {
  items: AcademicYearListChoiceItem[];
  current_start_year: number;
}

// =============================================================================
// Grade advance
// =============================================================================

export interface GradeAdvancePreviewItem {
  student_id: number;
  full_name: string;
  current_grade_level: number | null;
  current_is_graduate: boolean;
  current_academic_year_id: number | null;
  current_academic_year_name: string | null;
  suggested_grade_level: number | null;
  suggested_is_graduate: boolean;
  requires_track: boolean;
  has_track: boolean;
  has_reservations: boolean;
  has_completed_progress: boolean;
  blocker_notes: string[];
}

export interface GradeAdvancePreviewResponse {
  students: GradeAdvancePreviewItem[];
  suggested_year_id: number | null;
  suggested_year_name: string | null;
}

export interface GradeAdvanceApplyItem {
  student_id: number;
  new_grade_level: number | null;
  new_is_graduate: boolean;
  new_track?: Track | null;
  new_graduate_mode?: GraduateMode | null;
}

export interface GradeAdvanceApplyBody {
  items: GradeAdvanceApplyItem[];
  target_academic_year_id?: number | null;
}

export interface GradeAdvanceApplyResult {
  advanced_count: number;
  skipped_invalid: string[];
  skipped_track_missing: string[];
  preserved_reservations_count: number;
}

export interface ResetProgramConfirmBody {
  confirm_full_name: string;
}

export interface ResetProgramResult {
  student_id: number;
  deleted_tasks: number;
  deleted_task_book_items: number;
  cleared_reservations: number;
  deleted_suggestion_feedback: number;
}

// =============================================================================
// CSV
// =============================================================================

export interface CsvParsedRow {
  row_num: number;
  full_name: string | null;
  email: string | null;
  grade_level: number | null;
  track: Track | null;
  is_graduate: boolean;
  graduate_mode: GraduateMode | null;
  is_valid: boolean;
  errors: string[];
  warnings: string[];
  raw: Record<string, unknown>;
}

export interface CsvPreviewResponse {
  rows: CsvParsedRow[];
  valid_count: number;
  invalid_count: number;
  header_errors: string[];
  total_rows: number;
}

export interface CsvCommitBody {
  csv_text: string;
}

export interface CsvCreatedStudent {
  row_num: number;
  student_id: number;
  full_name: string;
  email: string;
  grade_label: string;
  temp_password: string;
}

export interface CsvCommitResult {
  created: CsvCreatedStudent[];
  skipped_existing_email: CsvParsedRow[];
  skipped_invalid: CsvParsedRow[];
  created_count: number;
  skipped_count: number;
  header_errors: string[];
}

File: app/routes/api_v2/schemas/academic.py
"""API v2 — Akademik yıl + dönem + sınıf yükseltme + CSV (Dalga 3 Paket 10).

Kapsam:
  - /api/v2/teacher/academic            : Akademik yıl + faz CRUD + öğrenci atama
  - /api/v2/teacher/grade-advance       : Toplu sınıf yükseltme preview + apply +
                                          tekil reset-program (çift onay)
  - /api/v2/teacher/csv                 : CSV import (preview→commit), CSV export

Referans Jinja (dokunulmaz):
  - app/routes/teacher_years.py
  - app/routes/teacher_students.py (promote + import akışları)
  - app/services/csv_import.py (parse + bulk_create_students)
"""
from __future__ import annotations

from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel


# =============================================================================
# Academic year + phase
# =============================================================================


PhaseKindLiteral = Literal[
    "regular", "winter_break", "summer_camp", "exam_prep",
]

ExamTargetLiteral = Literal["lgs", "yks", "none"]


class PhaseItem(BaseModel):
    id: int
    name: str
    start_date: date
    end_date: date
    kind: PhaseKindLiteral
    kind_label: str
    kind_badge: str
    notes: str | None
    capacity_multiplier: float
    is_no_school: bool


class AcademicYearListItem(BaseModel):
    id: int
    name: str
    start_year: int | None
    exam_target: ExamTargetLiteral
    exam_label: str
    is_active: bool
    phase_count: int
    student_count: int
    created_at: datetime


class AcademicYearListResponse(BaseModel):
    items: list[AcademicYearListItem]
    current_start_year: int          # Eylül-Ağustos ekseninde bugünün yılı


class AcademicYearAssignedStudent(BaseModel):
    student_id: int
    full_name: str
    grade_level: int | None
    is_graduate: bool


class AcademicYearDetailResponse(BaseModel):
    id: int
    name: str
    start_year: int | None
    exam_target: ExamTargetLiteral
    exam_label: str
    is_active: bool
    created_at: datetime
    phases: list[PhaseItem]
    assigned_students: list[AcademicYearAssignedStudent]


class AcademicYearCreateBody(BaseModel):
    """name otomatik 'YYYY-(YYYY+1)' üretilir; start_year < 2020 veya > 2050 → 422."""
    start_year: int                  # 2020..2050


class AcademicYearPatchBody(BaseModel):
    name: str | None = None
    start_year: int | None = None
    exam_target: ExamTargetLiteral | None = None
    is_active: bool | None = None


class PhaseCreateBody(BaseModel):
    name: str
    start_date: date
    end_date: date
    kind: PhaseKindLiteral = "regular"
    notes: str | None = None


class PhasePatchBody(BaseModel):
    name: str | None = None
    start_date: date | None = None
    end_date: date | None = None
    kind: PhaseKindLiteral | None = None
    notes: str | None = None


class AcademicYearAssignBody(BaseModel):
    """Diff atama:
    - Listede olup şu an başka yılda olan öğrenciler bu yıla taşınır
    - Listede olmayan ama şu an bu yılda olan öğrenciler academic_year_id=NULL
    - Zaten bu yılda olanlar değişmez
    """
    student_ids: list[int]


class AcademicYearAssignResult(BaseModel):
    assigned_count: int              # bu yıla yeni eklendi/değişti
    removed_count: int               # bu yıldan çıkarıldı (NULL'a düştü)
    unchanged_count: int


class AcademicYearListChoiceItem(BaseModel):
    """Yıl seçim listesi — UI'da hızlı 'şu yılı oluştur' butonu için."""
    start_year: int
    name: str
    label: str                       # "2026-2027 · şu an"
    exists: bool


class AcademicYearChoicesResponse(BaseModel):
    items: list[AcademicYearListChoiceItem]
    current_start_year: int


# =============================================================================
# Grade advance — toplu sınıf yükseltme + tekil reset-program
# =============================================================================


GraduateModeLiteral = Literal["full_time", "dershane"]
TrackLiteral = Literal["sayisal", "ea", "sozel", "dil"]


class GradeAdvancePreviewItem(BaseModel):
    student_id: int
    full_name: str
    current_grade_level: int | None
    current_is_graduate: bool
    current_academic_year_id: int | None
    current_academic_year_name: str | None
    suggested_grade_level: int | None
    suggested_is_graduate: bool
    requires_track: bool
    has_track: bool
    has_reservations: bool           # SectionProgress.reserved_count > 0
    has_completed_progress: bool     # SectionProgress.completed_count > 0
    blocker_notes: list[str]         # UI'da uyarı şeritleri


class GradeAdvancePreviewResponse(BaseModel):
    students: list[GradeAdvancePreviewItem]
    suggested_year_id: int | None    # Bir sonraki AY (varsa)
    suggested_year_name: str | None


class GradeAdvanceApplyItem(BaseModel):
    student_id: int
    new_grade_level: int | None      # None ile mezun
    new_is_graduate: bool
    new_track: TrackLiteral | None = None
    new_graduate_mode: GraduateModeLiteral | None = None


class GradeAdvanceApplyBody(BaseModel):
    items: list[GradeAdvanceApplyItem]
    target_academic_year_id: int | None = None   # tüm öğrenciler için (set/None)


class GradeAdvanceApplyResult(BaseModel):
    advanced_count: int
    skipped_invalid: list[str]                    # öğrenci adı + sebep
    skipped_track_missing: list[str]
    preserved_reservations_count: int             # değişmedi (rezerv koruma garantisi)


class ResetProgramConfirmBody(BaseModel):
    """Geri dönülemez veri kaybı — full_name yazılarak çift onay gelir."""
    confirm_full_name: str           # öğrenci.full_name ile birebir eşleşmeli


class ResetProgramResult(BaseModel):
    student_id: int
    deleted_tasks: int
    deleted_task_book_items: int
    cleared_reservations: int
    deleted_suggestion_feedback: int


# =============================================================================
# CSV import (preview → commit) + export
# =============================================================================


class CsvParseError(BaseModel):
    row_num: int                     # 1-indexed (header hariç)
    field: str | None = None
    message: str


class CsvParsedRow(BaseModel):
    row_num: int
    full_name: str | None
    email: str | None
    grade_level: int | None
    track: TrackLiteral | None
    is_graduate: bool
    graduate_mode: GraduateModeLiteral | None
    is_valid: bool
    errors: list[str]
    warnings: list[str]
    raw: dict


class CsvPreviewResponse(BaseModel):
    rows: list[CsvParsedRow]
    valid_count: int
    invalid_count: int
    header_errors: list[str]         # fatal — commit edilemez
    total_rows: int


class CsvCommitBody(BaseModel):
    """Önizleme sırasında validate edilen CSV metni aynen gönderilir.

    Server commit aşamasında YENİDEN parse + validate eder (tampered önler);
    sadece valid satırlar yaratılır.
    """
    csv_text: str


class CsvCreatedStudent(BaseModel):
    row_num: int
    student_id: int
    full_name: str
    email: str
    grade_label: str
    temp_password: str               # Tek seferlik — UI gösterimi sonrası unutulur


class CsvCommitResult(BaseModel):
    created: list[CsvCreatedStudent]
    skipped_existing_email: list[CsvParsedRow]
    skipped_invalid: list[CsvParsedRow]
    created_count: int
    skipped_count: int
    header_errors: list[str]         # fatal (kuota vs)
